AJC throws NPE when number of classpath entries is over the limit

1. Java 11 (AdoptOpenJDK jdk-11.0.10.9-hotspot)
2. aspectj-maven-plugin v1.14
3. aspectjtools v1.9.7
4. pom.xml snippet
<plugin>
   <groupId>org.codehaus.mojo</groupId>
   <artifactId>aspectj-maven-plugin</artifactId>
   <version>1.14</version>
   <configuration>
      <complianceLevel>11</complianceLevel>
      <source>11</source>
      <target>11</target>
      <showWeaveInfo>true</showWeaveInfo>
      <verbose>true</verbose>
      <aspectLibraries>
         <aspectLibrary>
            <groupId>xxx.xxx.xxx</groupId>
            <artifactId>xxx-aspects</artifactId>
         </aspectLibrary>
      </aspectLibraries>
      <includes>
         <include>**/*.aj</include>
         <include>**/*.class</include>
      </includes>
      <Xlint>cantFindType=ignore</Xlint>
      <weaveDirectories>
         <weaveDirectory>${project.build.directory}/classesToWeav</weaveDirectory>
      </weaveDirectories>
   </configuration>
   <executions>
      <execution>
         <!-- Compile and weave aspects after all classes compiled by javac -->
         <phase>process-classes</phase>
         <goals>
            <goal>compile</goal>
         </goals>
      </execution>
   </executions>
         <dependencies>
      <dependency>
         <groupId>org.aspectj</groupId>
         <artifactId>aspectjtools</artifactId>
         <version>1.9.7<version>
         <scope>compile</scope>
      </dependency>
   </dependencies>
</plugin>
 
5. Stack trace:
[ERROR] Failed to execute goal org.codehaus.mojo:aspectj-maven-plugin:1.14.0:compile (default) on project xxx: AJC compiler errors:
[ERROR] abort ABORT -- (NullPointerException) null
[ERROR] null
[ERROR] java.lang.NullPointerException
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.batch.ClasspathJmod.getModulesDeclaringPackage(ClasspathJmod.java:146)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.batch.ClasspathLocation.isPackage(ClasspathLocation.java:184)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.batch.ClasspathJmod.findClass(ClasspathJmod.java:55)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.batch.FileSystem.internalFindClass(FileSystem.java:544)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.batch.FileSystem.findClass(FileSystem.java:464)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.batch.FileSystem.findType(FileSystem.java:617)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.env.IModuleAwareNameEnvironment.findType(IModuleAwareNameEnvironment.java:101)
[ERROR]         at org.aspectj.ajdt.internal.core.builder.StatefulNameEnvironment.findType(StatefulNameEnvironment.java:101)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.createPlainPackage(LookupEnvironment.java:1151)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.buildTypeBindings(CompilationUnitScope.java:136)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.buildTypeBindings(LookupEnvironment.java:487)
[ERROR]         at org.aspectj.ajdt.internal.compiler.lookup.AjLookupEnvironment.buildTypeBindings(AjLookupEnvironment.java:1471)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.internalBeginToCompile(Compiler.java:870)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.beginToCompile(Compiler.java:395)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:449)
[ERROR]         at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:427)
[ERROR]         at org.aspectj.ajdt.internal.core.builder.AjBuildManager.performCompilation(AjBuildManager.java:1096)
[ERROR]         at org.aspectj.ajdt.internal.core.builder.AjBuildManager.performBuild(AjBuildManager.java:275)
[ERROR]         at org.aspectj.ajdt.internal.core.builder.AjBuildManager.batchBuild(AjBuildManager.java:188)
[ERROR]         at org.aspectj.ajdt.ajc.AjdtCommand.doCommand(AjdtCommand.java:103)
[ERROR]         at org.aspectj.ajdt.ajc.AjdtCommand.runCommand(AjdtCommand.java:47)
[ERROR]         at org.aspectj.tools.ajc.Main.run(Main.java:372)
[ERROR]         at org.aspectj.tools.ajc.Main.runMain(Main.java:250)
[ERROR]         at org.codehaus.mojo.aspectj.AbstractAjcCompiler.execute(AbstractAjcCompiler.java:568)
[ERROR]         at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR]         at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR]         at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR]         at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR]         at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR]         at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR]         at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR]         at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR]         at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR]         at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR]         at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR]         at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
[ERROR]         at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
[ERROR]         at org.apache.maven.cli.MavenCli.main(MavenCli.java:192)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
[ERROR]         at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[ERROR]

The problem is that ClasspathJmod objects for JMOD files have zipFile=null
 
![image](https://user-images.githubusercontent.com/100023594/154798321-3b1299ba-7d93-4205-9604-f60ef717e76a.png)

Problem can be solved by providing system property -Dorg.aspectj.weaver.openarchives=2000 when running maven build. It turns out due to big number of classpath entries some of these entries are closed during processing:
 
org.aspectj.org.eclipse.jdt.internal.compiler.batch.ClasspathJar
 
// AspectJ Extension
private void ensureOpen() throws IOException { //line 403
    if (zipFile != null) return; // If its not null, the zip is already open
    if (openArchives.size()>=maxOpenArchives) {
        closeSomeArchives(openArchives.size()/10); // Close 10% of those open <-- closed and likely not re-initialized so that zipFile remains null
    }
    zipFile = new ZipFile(file);
    openArchives.add(this);
}

Hi,

Classpath related NPE exception is fixed with this 1.9.9-SNAPSHOT release, however, I’m still getting the same AJC compiler error

[ERROR] Failed to execute goal org.codehaus.mojo:aspectj-maven-plugin:1.14.0:compile (default) on project xxx-app: AJC compiler errors:
[ERROR] error at (no source information available)
[ERROR] C:\Users\....aj:0::0 Type java.lang.Object is indirectly referenced from required .class files but cannot be resolved since the declaring package java.lang exported from module java.base conflicts with a package accessible from module <unnamed>

thus 1.9.4 has to be used.

Regards

From: Alexander Kriegisch 
Sent: Wednesday, February 23, 2022 5:04 PM
To: eclipse/org.aspectj 
Cc: raimondas67 ; Mention 
Subject: Re: [eclipse/org.aspectj] AJC throws NPE when number of classpath entries is over the limit (Issue #125)

@aclement, I inspected the existing logic and think it is safe to continue using the list instead of a set, mainly because it is more predictable and chronological which classpath JARs get closed on cache exhaustion (first in, first out). We would have to use a sorted set or map with a sortable timestamp or so as the key, which would entail more changes than I want to implement now. But I think I was pretty thorough with where to put the safeguards, my checks also encompassing subclasses. I think that for now we are OK, at least no worse than before. If @raimondas67 also re-tests successfully, I think the two PRs for JDT Core and AspectJ are ready to be reviewed and merged, whenever you have cycles to look into them.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android. 
You are receiving this because you were mentioned.Message ID: ***@***.***>

--- a/pom.xml
+++ b/pom.xml
@@ -21,7 +21,7 @@
 		<maven.javadoc.skip>true</maven.javadoc.skip>
 
 		<!-- Dependency versions -->
-		<jdt.core.version>1.9.8.RC3</jdt.core.version>
+		<jdt.core.version>1.9.9-SNAPSHOT</jdt.core.version>
 		<asm.version>9.2</asm.version>
 		<lib.ant.version>1.6.3</lib.ant.version>
 		<lib.ant.xerces.version>2.6.2</lib.ant.xerces.version>

--- a/tests/bugs198/github_125/Application.java
+++ b/tests/bugs198/github_125/Application.java
@@ -1,0 +1,11 @@
+public class Application {
+  public static void main(String[] argv) {
+    System.out.println("Hello world!");
+  }
+
+  static aspect MyAspect {
+    before(): execution(* Application.main(..)) {
+      System.out.println("Before advice");
+    }
+  }
+}

--- a/tests/pom.xml
+++ b/tests/pom.xml
@@ -90,6 +90,17 @@
 			<version>${project.version}</version>
 			<scope>test</scope>
 		</dependency>
+		<!--
+			Used for tests setting system properties, e.g. for AJC, when we need to avoid polluting the global
+			system properties namespace with settings potentially influencing other tests.
+			See https://github.com/bmuskalla/scoped-system-properties.
+		-->
+		<dependency>
+			<groupId>io.github.bmuskalla</groupId>
+			<artifactId>scoped-system-properties</artifactId>
+			<version>1.1.0</version>
+			<scope>test</scope>
+		</dependency>
 
 		<!--
 			The tests need these during runtime, even though no direct usage is in our classes.

--- a/tests/src/test/java/org/aspectj/systemtest/ajc198/Ajc198TestsJava.java
+++ b/tests/src/test/java/org/aspectj/systemtest/ajc198/Ajc198TestsJava.java
@@ -39,26 +39,6 @@ public class Ajc198TestsJava extends XMLBasedAjcTestCaseForJava17OrLater {
 		checkVersion("PersonAspect", Constants.MAJOR_17, 0 /* Constants.PREVIEW_MINOR_VERSION */);
 	}
 
-	public void testAnnotationStyleSpecialIfClauses() {
-		runTest("annotation style A");
-	}
-
-	public void testAnnotationStylePointcutInheritanceWithIfClauses() {
-		runTest("annotation style B");
-	}
-
-	public void testAnnotationStyleSpecialIfClauses2_gh120() {
-		runTest("annotation style C");
-	}
-
-	public void testAnnotationStyleSpecialIfClauses3_gh120() {
-		runTest("annotation style D");
-	}
-
-	public void testAnnotationStyleNegatedIf_gh122() {
-		runTest("annotation style negated if");
-	}
-
 	public static Test suite() {
 		return XMLBasedAjcTestCase.loadSuite(Ajc198TestsJava.class);
 	}

--- a/tests/src/test/java/org/aspectj/systemtest/ajc198/Bugs198Tests.java
+++ b/tests/src/test/java/org/aspectj/systemtest/ajc198/Bugs198Tests.java
@@ -7,6 +7,8 @@
  *******************************************************************************/
 package org.aspectj.systemtest.ajc198;
 
+import io.bmuskalla.system.properties.PropertyEnvironment;
+import io.bmuskalla.system.properties.ScopedSystemProperties;
 import junit.framework.Test;
 import org.aspectj.testing.XMLBasedAjcTestCase;
 
@@ -15,9 +17,9 @@ import org.aspectj.testing.XMLBasedAjcTestCase;
  */
 public class Bugs198Tests extends XMLBasedAjcTestCase {
 
-	public void testGitHub_105() {
-		runTest("ITD annotation with mandatory parameter via aspectpath");
-	}
+  public void testGitHub_105() {
+    runTest("ITD annotation with mandatory parameter via aspectpath");
+  }
 
   public void testAnnotationStyleSpecialIfClauses() {
     runTest("annotation style A");
@@ -37,6 +39,13 @@ public class Bugs198Tests extends XMLBasedAjcTestCase {
 
   public void testAnnotationStyleNegatedIf_gh122() {
     runTest("annotation style negated if");
+  }
+
+  public void testGitHub_125() {
+    try (PropertyEnvironment env = ScopedSystemProperties.newPropertyEnvironment()) {
+      env.setProperty("org.aspectj.weaver.openarchives", "20");
+      runTest("compiler can re-open closed JARs");
+    }
   }
 
   public static Test suite() {

--- a/tests/src/test/java/org/aspectj/systemtest/ajc198/Bugs198Tests.java
+++ b/tests/src/test/java/org/aspectj/systemtest/ajc198/Bugs198Tests.java
@@ -19,6 +19,26 @@ public class Bugs198Tests extends XMLBasedAjcTestCase {
 		runTest("ITD annotation with mandatory parameter via aspectpath");
 	}
 
+  public void testAnnotationStyleSpecialIfClauses() {
+    runTest("annotation style A");
+  }
+
+  public void testAnnotationStylePointcutInheritanceWithIfClauses() {
+    runTest("annotation style B");
+  }
+
+  public void testAnnotationStyleSpecialIfClauses2_gh120() {
+    runTest("annotation style C");
+  }
+
+  public void testAnnotationStyleSpecialIfClauses3_gh120() {
+    runTest("annotation style D");
+  }
+
+  public void testAnnotationStyleNegatedIf_gh122() {
+    runTest("annotation style negated if");
+  }
+
   public static Test suite() {
     return XMLBasedAjcTestCase.loadSuite(Bugs198Tests.class);
   }

--- a/tests/src/test/resources/org/aspectj/systemtest/ajc198/ajc198.xml
+++ b/tests/src/test/resources/org/aspectj/systemtest/ajc198/ajc198.xml
@@ -153,4 +153,18 @@
 		</run>
 	</ajc-test>
 
+	<ajc-test dir="bugs198/github_125" title="compiler can re-open closed JARs">
+		<!--
+			Here the Java test sets system property org.aspectj.weaver.openarchives to 20 in order to provoke
+			open JAR limit exhaustion
+		-->
+		<compile files="Application.java" options="-1.5" />
+		<run class="Application">
+			<stdout>
+				<line text="Before advice"/>
+				<line text="Hello world!"/>
+			</stdout>
+		</run>
+	</ajc-test>
+
 </suite>

